File: src/bun/accounts.ts
import { createHash, randomBytes } from "node:crypto";
import type { Db } from "./db";
import type { UserDto } from "../shared/types";

export interface UserRow {
  id: string; username: string; display_name: string; bio: string;
  password_hash: string | null; is_private: number; created_at: string;
}
export const publicUser = (u: UserRow): UserDto => ({
  id: u.id, username: u.username, displayName: u.display_name, bio: u.bio,
  isPrivate: !!u.is_private, isGuest: !u.password_hash, createdAt: u.created_at,
});
const digest = (token: string) => createHash("sha256").update(token).digest("hex");
export function accounts({ db }: Db) {
  const byId = (id: string) => db.query<UserRow, [string]>("SELECT * FROM users WHERE id = ?").get(id);
  return {
    byId,
    byUsername: (name: string) => db.query<UserRow, [string]>("SELECT * FROM users WHERE username = ? COLLATE NOCASE AND password_hash IS NOT NULL").get(name),
    authenticate: (authorization: string | null) => {
      if (!authorization?.startsWith("Bearer ")) return null;
      return db.query<UserRow, [string, number]>("SELECT u.* FROM users u JOIN auth_tokens t ON t.user_id = u.id WHERE t.token_hash = ? AND t.expires_at > ?").get(digest(authorization.slice(7)), Date.now());
    },
    issue: (user: UserRow) => {
      const token = randomBytes(32).toString("hex");
      db.query("DELETE FROM auth_tokens WHERE expires_at <= ?").run(Date.now());
      db.query("INSERT INTO auth_tokens VALUES (?, ?, ?)").run(digest(token), user.id, Date.now() + 30 * 86400000);
      return { token, user: publicUser(user) };
    },
    revoke: (authorization: string | null) => {
      if (authorization?.startsWith("Bearer ")) db.query("DELETE FROM auth_tokens WHERE token_hash = ?").run(digest(authorization.slice(7)));
    },
    guest: () => {
      const id = crypto.randomUUID();
      db.query("INSERT INTO users (id, username, display_name) VALUES (?, ?, 'Guest')").run(id, `guest-${id}`);
      return byId(id)!;
    },
    register: (username: string, displayName: string, hash: string, guestId?: string) => db.transaction(() => {
      const id = guestId ?? crypto.randomUUID();
      if (guestId) {
        db.query("UPDATE users SET username = ?, display_name = ?, password_hash = ? WHERE id = ? AND password_hash IS NULL").run(username, displayName, hash, id);
        // Rotate every anonymous token when the guest becomes a password-protected account.
        db.query("DELETE FROM auth_tokens WHERE user_id = ?").run(id);
      } else db.query("INSERT INTO users (id, username, display_name, password_hash) VALUES (?, ?, ?, ?)").run(id, username, displayName, hash);
      return byId(id)!;
    })(),
    update: (id: string, displayName: string, bio: string, isPrivate: boolean) => {
      db.query("UPDATE users SET display_name = ?, bio = ?, is_private = ? WHERE id = ?").run(displayName, bio, Number(isPrivate), id);
      return publicUser(byId(id)!);
    },
    search: (query: string) => db.query<UserRow, [string, string]>(
      "SELECT * FROM users WHERE password_hash IS NOT NULL AND is_private = 0 AND (instr(lower(username), ?) > 0 OR instr(lower(display_name), ?) > 0) ORDER BY username LIMIT 30"
    ).all(query.toLowerCase(), query.toLowerCase()).map(publicUser),
  };
}

File: rust-api/src/accounts.rs
use crate::{
    db::{one, required},
    error::{ApiError, Result},
};
use rand::RngCore;
use rusqlite::{Connection, params};
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use std::time::{SystemTime, UNIX_EPOCH};
pub fn now() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}
pub fn digest(token: &str) -> String {
    format!("{:x}", Sha256::digest(token.as_bytes()))
}
pub fn public(user: &Value) -> Value {
    json!({"id":user["id"],"username":user["username"],"bio":user["bio"],"isGuest":user["password_hash"].is_null(),"createdAt":user["created_at"]})
}
pub fn auth(db: &Connection, token: &str) -> Result<Option<Value>> {
    let Some(token) = token.strip_prefix("Bearer ") else {
        return Ok(None);
    };
    one(
        db,
        "SELECT u.* FROM users u JOIN auth_tokens t ON t.user_id=u.id WHERE t.token_hash=? AND t.expires_at>?",
        params![digest(token), now()],
    )
}
pub fn signed_in(db: &Connection, token: &str) -> Result<Value> {
    auth(db, token)?.ok_or_else(|| ApiError::new(401, "Please sign in again."))
}
pub fn by_username(db: &Connection, name: &str) -> Result<Option<Value>> {
    one(
        db,
        "SELECT * FROM users WHERE username=? COLLATE NOCASE AND password_hash IS NOT NULL",
        [name],
    )
}
pub fn issue(db: &Connection, user: &Value) -> Result<Value> {
    let mut bytes = [0u8; 32];
    rand::rngs::OsRng.fill_bytes(&mut bytes);
    let token: String = bytes.iter().map(|b| format!("{b:02x}")).collect();
    db.execute("DELETE FROM auth_tokens WHERE expires_at<=?", [now()])?;
    db.execute(
        "INSERT INTO auth_tokens VALUES(?,?,?)",
        params![digest(&token), user["id"].as_str(), now() + 30 * 86400000],
    )?;
    Ok(json!({"token":token,"user":public(user)}))
}
pub fn guest(db: &Connection) -> Result<Value> {
    let id = uuid::Uuid::new_v4().to_string();
    db.execute(
        "INSERT INTO users(id,username) VALUES(?,?)",
        params![id, format!("guest-{id}")],
    )?;
    let user = required(
        db,
        "SELECT * FROM users WHERE id=?",
        [id],
        "Unknown account",
    )?;
    issue(db, &user)
}
pub fn register(
    db: &mut Connection,
    username: &str,
    hash: &str,
    guest_id: Option<String>,
) -> Result<Value> {
    let tx = db.transaction()?;
    if by_username(&tx, username)?.is_some() {
        return Err(ApiError::new(409, "This username is already taken."));
    }
    let id = guest_id
        .clone()
        .unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
    if guest_id.is_some() {
        if tx.execute(
            "UPDATE users SET username=?,password_hash=? WHERE id=? AND password_hash IS NULL",
            params![username, hash, id],
        )? != 1
        {
            return Err(ApiError::new(
                409,
                "This guest already has an account. Sign in.",
            ));
        }
        tx.execute("DELETE FROM auth_tokens WHERE user_id=?", [&id])?;
    } else {
        tx.execute(
            "INSERT INTO users(id,username,password_hash) VALUES(?,?,?)",
            params![id, username, hash],
        )?;
    }
    let user = required(
        &tx,
        "SELECT * FROM users WHERE id=?",
        [id],
        "Unknown account",
    )?;
    let response = issue(&tx, &user)?;
    tx.commit()?;
    Ok(response)
}
